### ISSUE TITLE
bottled package installs broken with non-standard install prefix

### ISSUE THREAD

**Issue by jvandew** [Homebrew/brew#13209]
### `brew config` output

```shell
HOMEBREW_VERSION: 3.4.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: d0a0bbef8db45cca303c607037338d02d8eaa8ca
Last commit: 4 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 89bb4ce3b7b86c85a8bd2d60db8fa829e8ff9c1a
Core tap last commit: 9 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /Users/jacob/homebrew
HOMEBREW_REPOSITORY: /Users/jacob/homebrew
HOMEBREW_CELLAR: /Users/jacob/homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /Users/jacob/homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: octa-core 64-bit westmere
Clang: 13.0.0 build 1300
Git: 2.32.0 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.6-x86_64
CLT: 13.2.0.0.1.1638488800
Xcode: 13.2.1
Rosetta 2: true
```

### `brew doctor` output

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Homebrew's prefix is not /usr/local.
Some of Homebrew's bottles (binary packages) can only be used with the default
prefix (/usr/local).
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
unsupported configuration.


Warning: Homebrew's "bin" was not found in your PATH.
Consider setting your PATH for example like so:
  echo 'export PATH="/Users/jacob/homebrew/bin:$PATH"' >> ~/.zshrc
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [x] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

We're currently managing local development environments via a bootstrapped homebrew install in a custom location, isolated from users' system brew install. To speed up required package installs, we use the `--force-bottle` flag to install bottled binaries rather than building from source.

### What happened (include all command output)?

Any install using `--force-bottle` fails with the following error:
```
~ ➤ homebrew/bin/brew install --force-bottle gettext                                                                             
==> Downloading https://ghcr.io/v2/homebrew/core/gettext/manifests/0.21
Already downloaded: /Users/jacob/Library/Caches/Homebrew/downloads/66569d4eb02ec67195446d3dcba7d7806a35f88f629a55b43ba53fb1febb1f56--gettext-0.21.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/gettext/blobs/sha256:a025e143fe3f5f7e24a936b8b0a4926acfdd025b11d62024e3d355c10653
Already downloaded: /Users/jacob/Library/Caches/Homebrew/downloads/56f8e2a346bf8ce27f5f2aadbc0b3163e78d8cae1fcfd9f462477cd4b821479f--gettext--0.21.big_sur.bottle.tar.gz
==> Pouring gettext--0.21.big_sur.bottle.tar.gz
Error: Patching failed!  Original and patched binary sizes do not match.
Original size: 68912
Patched size: 68923 
```
Running with `--debug` additionally spits out a bunch of messages along the lines of
```
==> Changing install name in /Users/jacob/homebrew/Cellar/gettext/0.21/bin/envsubst
  from @@HOMEBREW_CELLAR@@/gettext/0.21/lib/libintl.8.dylib
    to /Users/jacob/homebrew/Cellar/gettext/0.21/lib/libintl.8.dylib
==> Changing install name in /Users/jacob/homebrew/Cellar/gettext/0.21/bin/gettext
  from @@HOMEBREW_CELLAR@@/gettext/0.21/lib/libintl.8.dylib
    to /Users/jacob/homebrew/Cellar/gettext/0.21/lib/libintl.8.dylib
==> Changing install name in /Users/jacob/homebrew/Cellar/gettext/0.21/bin/msgattrib
  from @@HOMEBREW_CELLAR@@/gettext/0.21/lib/libgettextsrc-0.21.dylib
    to /Users/jacob/homebrew/Cellar/gettext/0.21/lib/libgettextsrc-0.21.dylib
```

### What did you expect to happen?

the package installs successfully just as it does with the default prefix:
```
~ ➤ which brew                                                                                                                   
/usr/local/bin/brew
~ ➤ brew reinstall --force-bottle gettext                                                                                        
==> Downloading https://ghcr.io/v2/homebrew/core/gettext/manifests/0.21
Already downloaded: /Users/jacob/Library/Caches/Homebrew/downloads/66569d4eb02ec67195446d3dcba7d7806a35f88f629a55b43ba53fb1febb1f56--gettext-0.21.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/gettext/blobs/sha256:a025e143fe3f5f7e24a936b8b0a4926acfdd025b11d62024e3d355c10653
Already downloaded: /Users/jacob/Library/Caches/Homebrew/downloads/56f8e2a346bf8ce27f5f2aadbc0b3163e78d8cae1fcfd9f462477cd4b821479f--gettext--0.21.big_sur.bottle.tar.gz
==> Reinstalling gettext 
==> Pouring gettext--0.21.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/gettext/0.21: 1,953 files, 19.8MB
==> Running `brew cleanup gettext`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
```

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
1. set up a new homebrew install in a custom location following the steps [here](https://github.com/Homebrew/brew/blob/664d0c67d5947605c914c4c56ebcfaa80cb6eca0/docs/Installation.md#untar-anywhere):

mkdir homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
```
2. use it to install a package with `--force-bottle`:
```
homebrew/bin/brew install --force-bottle gettext
```
```

**Comment by jvandew** [comment]
Digging around, it looks like this is possibly related to #13138 which landed recently? Installing packages from source seems to be unaffected, so for the time being we're just omitting the `--force-bottle` flag as a workaround.

**Comment by carlocab** [comment]
CC @danielnachun

Just FYI, @jvandew: we don't provide support for non-default prefixes since there are just too many things that can go wrong with them, particularly if you use the `--force-bottle` flag [1]. I'll leave this open for now but I suspect another maintainer will close this issue soon.

[1] The `--force-bottle` flag overrides the relocatability check, and will install a bottle even when we have strong reason to believe that the bottle won't work, so having problems with it is expected.

**Comment by danielnachun** [comment]
The reason this bug is happening is because the new patching code we've added will try to replace prefix references with the new prefix in binaries.  It is not supposed to run unless you've set `HOMEBREW_RELOCATE_BUILD_PREFIX` to a non-empty value and your non-default prefix is shorter than the default prefix for your platform (in your case `/usr/local`), but I didn't consider the scenario where `--force-bottle` is used.  This is a simple fix I will make ASAP as this shouldn't be happening.

On a related note, if you are able to use a prefix that is shorter than `/usr/local` (`/opt/hb` is the one I've used for a lot of testing) and you set `HOMEBREW_RELOCATE_BUILD_PREFIX`, you don't need to use `--force-bottle` anymore and **most** bottles will work without issue.  This is still a highly experimental feature and there are definitely bottles that don't work, but it's moving us towards greatly reducing the need to build from source for non-default prefixes.  

I'm okay with keeping this open until I've fixed this as I do consider this a bug, but I do understand if other maintainers want to close it.

**Comment by jvandew** [comment]
Awesome, thank you for the additional context and looking into a fix. Let me know if there is anything else I can do to help.

**Comment by danielnachun** [comment]
I've opened https://github.com/Homebrew/brew/pull/13217 to fix this.  I tested it using `--force-bottle` without setting `HOMEBREW_RELOCATE_BUILD_PREFIX` and it did not attempt to do any patching.

**Comment by carlocab** [comment]
Fixed in #13217.

**Comment by jvandew** [comment]
`brew developer on && brew update` didn't seem to pick up the fix for me, but reinstalling via https://github.com/Homebrew/brew/tarball/master did the trick.

Thanks for quick turnaround!
